### PR TITLE
api: Add methods to join components using a collector

### PR DESCRIPTION
**Pull request by zml2008**
Closes GH-286

### CODE REVIEW

File: api/src/main/java/net/kyori/adventure/text/ComponentBuilder.java [hunk 1]
```diff
@@ -155,6 +156,14 @@
   @Contract("_ -> this")
   @NonNull B mapChildrenDeep(final @NonNull Function<BuildableComponent<?, ?>, ? extends BuildableComponent<?, ?>> function);
 
+  /**
+   * Get an unmodifiable list containing all children currently in this builder.
+   *
+   * @return the list of children
+   * @since 4.6.0
+   */
+  @NonNull List<Component> children();
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/ComponentBuilder.java:165]: Is this method necessary?

Comment (zml2008) [reply api/src/main/java/net/kyori/adventure/text/ComponentBuilder.java:165]: it's not strictly required -- the collector could just work with the builder implementation -- but it feels weird to not expose functionality the collector needs to the api.

File: api/src/test/java/net/kyori/adventure/text/AbstractComponentTest.java [hunk 2]
```diff
@@ -376,4 +377,40 @@ void testBuilderApplyDeep() {
...
+      .append(Component.space())
+      .append(Component.text("World"))
+      .build();
+
+    assertEquals(expected, joined);
+
+  }
+
+  @Test
+  void testJoinEmpty() {
```
Comment (kashike) [inline api/src/test/java/net/kyori/adventure/text/AbstractComponentTest.java:413]:
```suggestion
  void testCollectOnEmptyStreamReturnsEmptyComponent() {
```

File: api/src/test/java/net/kyori/adventure/text/AbstractComponentTest.java [hunk 3]
```diff
@@ -376,4 +377,40 @@ void testBuilderApplyDeep() {
...
+    ).collect(Component.toComponent(Component.space()));
+
+    final Component expected = Component.text()
+      .append(Component.text("Hello", NamedTextColor.RED))
+      .append(Component.space())
+      .append(Component.text("World"))
+      .build();
+
+    assertEquals(expected, joined);
+
```
Comment (kashike) [inline api/src/test/java/net/kyori/adventure/text/AbstractComponentTest.java:409]:
```suggestion
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: api/src/main/java/net/kyori/adventure/text/ComponentBuilder.java (hunk 1 above)
Comment (kashike) [reply api/src/main/java/net/kyori/adventure/text/ComponentBuilder.java:165]: I'm not really a fan of exposing getters on builders... but it looks like we do for `content()` as well, so I guess this is okay.

**Review by kashike — APPROVED** [review]

**Pull request merged**